src/main/assisted-installer-controller/assisted_installer_main.go

```diff
@@ -55,38 +62,77 @@ func main() {
 		kc,
 	)
 
+	assistedController.SetReadyState()
+
 	// While adding new routine don't miss to add wg.add(1)
 	// without adding it will panic
 	var wg sync.WaitGroup
-	var wgLogs sync.WaitGroup
 	var status assistedinstallercontroller.ControllerStatus
 
-	ctxApprove, cancelApprove := context.WithCancel(context.Background())
-	go assistedController.ApproveCsrs(ctxApprove, &wg)
+	ctxRoutines, cancelRoutines := context.WithCancel(context.Background())
+	go assistedController.WaitAndUpdateNodesStatus(ctxRoutines, &wg)
 	wg.Add(1)
-	go assistedController.PostInstallConfigs(&wg, &status)
+	go assistedController.PostInstallConfigs(ctxRoutines, &wg, &status)
 	wg.Add(1)
-	go assistedController.UpdateBMHs(&wg)
+	go assistedController.UpdateBMHs(ctxRoutines, &wg)
 	wg.Add(1)
+
+	// No need to cancel with context, will finish quickly
 	go assistedController.HackDNSAddressConflict(&wg)
 	wg.Add(1)
 
-	ctxLogs, cancelLogs := context.WithCancel(context.Background())
-	go assistedController.UploadLogs(ctxLogs, cancelLogs, &wgLogs, &status)
-	wgLogs.Add(1)
+	go assistedController.UploadLogs(ctxRoutines, &wg, &status)
+	wg.Add(1)
 
-	assistedController.SetReadyState()
-	assistedController.WaitAndUpdateNodesStatus(&status)
-	logger.Infof("Sleeping for 10 minutes to give a chance to approve all csrs")
-	time.Sleep(10 * time.Minute)
-	cancelApprove()
+	waitForInstallation(client, logger, &status)
```

Comment:
I strongly prefer the listener option, but if you implement this this way at least document that returning from this function means that we either completed or aborted and there is nothing that should come between this routine an cancel context

Reply:
i think it is very clear from the code

Reply:
I'm afraid it is really not (and I actually knew the code before reviewing it)

Reply:
Not sure i understand the complexity of it but i will document